### PR TITLE
[1 of 2] Refactor EYB lead schema

### PR DESCRIPTION
**Pull request by oliverjwroberts**
### Description of change

This PR is 1 of 2 that works toward ingesting EYB data into Data Hub, and is going into a feature branch. The overall plan is to ingest EYB leads, stored as JSON in Data Hub's S3 bucket, into Data Hub via the company activity jobs/ingestion infrastructure.

Notable changes include:
- Removal of the EYB lead create view/endpoint as we are no longer going to ingest leads this way.
- Sector data will be received as four fields (`sectorID`, `sector`, `sectorSub`, `sectorSubSub`), but stored as a `sector` field that contains the related instance and `sector_segments` which contains the individual segment names; we cannot use `sectorID` to lookup a sector as this is the `SLXXX` ID from the reference Data Workspace dataset, which we do not store. Instead, we have to use the segment names.
- Additional company address fields will now be received, including `companyLocation` which contains a country's ISO Alpha 2 code (and is subsequently stored as `address_country` containing the specified country instance
- We also will get an optional `dunsNumber` to help us identify the specific company, when provided.

I have opted to use Serialisers for each _group_ of fields (triage and user) to provide validation of the incoming JSON and translate the data to Python types/an EYBLead instance (as a lot of the functionality comes straight out of the box). This is instead of going for a dictionary of mappings approach. 

At the moment, these Serialiser's aren't being used, but that will come in part 2 of the work where we shall define the job to actually ingest the JSON.

### camelCase fields in the Serialiser
As it currently stands, I have opted to use `camelCase` fields within the `CreateEYBLead` serialisers because the incoming fields will be camelCase. I think this makes the code more readable, but appreciate it goes against Python convention (and flake8 was not happy either). 

As a feasibility test, I have successfully removed the need for `camelCase` serialiser fields by adding a _camel-to-snake_ function in the `to_internal_value` method, among other changes:

```python
    def to_internal_value(self, data):
        """Convert unvalidated JSON data to validated data."""
        snake_case_data = {}   <------
        for key, value in data.items():    <------
            mapped_key = inflection.underscore(key)    <------
            snake_case_data[mapped_key] = value    <------
        internal_value = super().to_internal_value(snake_case_data)
        related_fields = self.get_related_fields_internal_value(data)
        internal_value.update(related_fields)
        return internal_value
```

This, however, has meant that throughout the data's lifecycle within the Serialisers, some fields can have up to 3 different names. For example, company location represents the country that an investing company originates from: in the incoming JSON, this field is `companyLocation`; in the validation functions of the serialiser, this field is `company_location`; and finally, in the serialiser's validated_data, the field is `address_country` to conform to the Address serialiser requirements. I felt this was another layer of complexity, just to make flake8 happy? However, I would like to get others' opinions on this, please.

> I did look at third party packages like Django REST Framework JSON CamelCase package](https://github.com/vbabiy/djangorestframework-camel-case), but this essentially does the same thing under the hood.

### Checklist

* [x] Has this branch been rebased on top of the current `main` branch?

  <details>
  <summary>Explanation</summary>
  
  The branch should not be stale or have conflicts at the time reviews are requested.
  
  </details>

* [ ] Is the CircleCI build passing?

### General points

<details>
<summary><strong>Other things to check</strong></summary><p></p>

* Make sure `fixtures/test_data.yaml` is maintained when updating models
* Consider the admin site when making changes to models
* Use select-/prefetch-related field lists in views and search apps, and update them when fields are added
* Make sure the README is updated e.g. when adding new environment variables

</details>

See [docs/CONTRIBUTING.md](https://github.com/uktrade/data-hub-api/blob/main/docs/CONTRIBUTING.md) for more guidelines.

### CODE REVIEW

File: datahub/investment_lead/test/test_services.py [hunk 1]
```diff
@@ -82,12 +82,13 @@ def test_add_new_company_without_company_name_fails(self):
...
-        eyb_lead.address_county = None
+    # TODO: fix this as it's now failing after a typo correction
+    # def test_add_new_company_without_address_country_fails(self):
+    #     eyb_lead = EYBLeadFactory(duns_number=None)
+    #     eyb_lead.address_country = None
 
-        with pytest.raises(IntegrityError):
-            add_new_company_from_eyb_lead(eyb_lead)
+    #     with pytest.raises(IntegrityError):
+    #         add_new_company_from_eyb_lead(eyb_lead)
```
Comment (marijnkampf) [inline datahub/investment_lead/test/test_services.py:92]: Should these comments be removed from this PR? Presuming this is resolved in PR 2/2.

File: datahub/investment_lead/serializers.py [hunk 2]
```diff
@@ -86,137 +76,319 @@
...
-    This serializer has been tailored to the data we expect to receive from EYB/DataFlow.
-    Most notably, this includes string representations of related model instances.
-    """
+    def get_related_fields_representation(self, instance):
+        """Provides related fields in a representation-friendly format; override as needed."""
+        return {}
+
+
+class CreateEYBLeadTriageSerializer(BaseEYBLeadSerializer):
+    """Serializer for creating an EYB lead from triage data."""
```
Comment (marijnkampf) [inline datahub/investment_lead/serializers.py:129]: I would add a short comment explaining why we use camelCase in the serialiser and that fields are marked with `# noqa: N815`.

File: datahub/investment_lead/serializers.py [hunk 3]
```diff
@@ -86,137 +76,319 @@
...
+            'sector': level_zero_segment,
+            'sectorSub': level_one_segment,
+            'sectorSubSub': level_two_segment,
+            'location': instance.proposed_investment_region.name
+            if instance.proposed_investment_region else None,
+        }
+
+
+class CreateEYBLeadUserSerializer(BaseEYBLeadSerializer):
+    """Serializer for creating an EYB lead from user data."""
```
Comment (marijnkampf) [inline datahub/investment_lead/serializers.py:274]: Same as above; I would add a short comment explaining why we use camelCase in the serialiser and that fields are marked with `# noqa: N815`.

**Review by marijnkampf — APPROVED** [review]

File: datahub/investment_lead/models.py [hunk 4]
```diff
@@ -78,65 +79,56 @@ class LandingTimeframeChoices(models.TextChoices):
...
+    )
     intent = ArrayField(
         models.CharField(max_length=CHAR_FIELD_MAX_LENGTH, choices=IntentChoices.choices),
         size=6,
         default=list,
     )
     intent_other = models.CharField(max_length=CHAR_FIELD_MAX_LENGTH, default='', blank=True)
-    location = models.ForeignKey(
+    # We receive location, locationCity, and locationNone from EYB;
+    # These fields represent the proposed location of investment (i.e. UK region etc);
```
Comment (samuele-mattiuzzo) [inline datahub/investment_lead/models.py:105]:
```suggestion
    # These fields represent the proposed location of investment (i.e. UK region etc)
```

**Review by samuele-mattiuzzo — APPROVED** [review]
left a small comment + I agree with @marijnkampf comments

File: datahub/investment_lead/test/test_services.py (hunk 1 above)
Comment (oliverjwroberts) [reply datahub/investment_lead/test/test_services.py]: This test previously set `eyb_lead.address_county` as None. I interpreted the test name to mean it should instead set `.address_country` as None. Due to `Company.address_country` being a nullable field (as it's a foreign key), it does NOT raise an IntegrityError. 

To address this, I've added an additional check in `add_new_company_from_eyb_lead` to raise a ValueError if the field is None. I hope that is satisfactory?

File: datahub/investment_lead/models.py (hunk 4 above)
Comment (oliverjwroberts) [reply datahub/investment_lead/models.py:105]: I added the punctuation because the individual comments are linked and I was writing them as bullet points.

**Pull request merged**